### ISSUE TITLE
GMT writes invalid setdash command for lines too thin

### ISSUE THREAD

**Issue by PaulWessel** [GenericMappingTools/gmt#6833]
**Description of the problem**

See this [forum post](https://forum.generic-mapping-tools.org/t/failed-to-plot-large-multiple-polygon-vector-shp-file-in-pygmt/3064/8) for an example - I have had similar ones in the past but not understood the issue fully.  Now I do.

The PSL library, in order to keep PostScript files reasonably small, uses a smallest increment unit of 1/1200 inches.  I.e., it rounds to an effective dpi of 1200.  This means any pen thickness is a multiple of 0.06 points. Note, the DPI in GMT 4 used to be 300. We therefore only write _integer_ PS units.

Except for a few places (including angles), all coordinates and pens are given in integers which are 1/1200 inches or 0.06 points.  If you specify a pen of 0.06p then PSL writes `1 W` to set a pen thickness of 1 PS unit.  Smaller pens gives `0 W`, a zero-width pen, which is what **-Wfaint** means.  So -W0.0000000001p will still work fine - it is by definition the thinnest line possible by the PostScript device (printer, gs, whatever).  Telling PostScript to draw a line with pen thickness 0 is thus valid.

However, when we set dashed lines we specify the length of each dash and the gaps between them.  We use the same units there.  So, -W0.001p,lightseagreen,- is means a dash of 8 units with a gap of 4 units, i.e., two lengths of 0.008p and 0.004p.  These, when converted to our PS unit yields 0.133333 and 0.066666 PS units.  Being below 0.06p they round to 0 and we end up with `[0 0 ] 0 B` which is not allowed by PostScript.

There are a few different solutions here:

1. We could add a check that if any of the dash or gap lengths are 0 PS units then we skip setting a dashed line since it would result in an error.
2. We could write the _setdash_ output using floating point.  Here, that means [0.1333333 0.06666667] 0 B, a valid setting. If the device is capable of this then we would see dashes.
3. We update our 1200 DPI constant to something bigger, e.g, 12000.

I dismiss the last one since this will bloat most plot files by a lot without any real improvement.  And it would not fix this problem, since we now have that gap being 0.666667 PS unit which is still 0 and error. I also dismiss the first one as being too cumbersome and will result in solid lines when a dashed one would actually be visible if you zoomed in enough.

My preferred solution is therefore to do the _setdash_ using floating points, as we do for angles.  This will prevent this problem from reoccurring. Also, no bloating will occur since this command only happens a few times for most plots, e.g., once per line.

**Comment by PaulWessel** [comment]
Doing this affects current plots a tiny bit, leading to ~35 failures since` [40 20] 5 B `now may look like `[39.78856 20.0454] 4.98997 B`.  I think that is worth it to void a more complicated fix (what would that be? Determine that the segment lenghts are "lon enough" to round to integers?).  Any thoughts from @joa-quim on this?

**Comment by joa-quim** [comment]
Nothing in particular. When is the decision taken to pass in floats to pslib? When otherwise it would round to zero?

**Comment by PaulWessel** [comment]
We always pass floats into PSL, but PSL writes integers, mostly (except as noted above).  The rounding happens in PSL.  So this is entirely limited to postscriptlight.c.

**Comment by joa-quim** [comment]
OK, but the question is the same. When will the floats be passed to the PS file, when rounding them with result into 0?

**Comment by joa-quim** [comment]
My question is if we keep writing integers in the postscript file, except when otherwise that would write zeros, or they will all become floats.

**Comment by PaulWessel** [comment]
No, this is restricted to set dash commends only, of which there are 0-n in any PostScript file, with n being a small integer.  So no practical effect on plot size.  But fixes a bug.

**Comment by PaulWessel** [comment]
I tried a slightly more elaborate fix that only leads to 3 failures:

```
			w = atof(pattern) * PSL->internal.dpp;
			if (w > 4.0)	/* Set as integer PS. Max error 12.5% (e.g., 4.499999 -> 4) */
				sprintf (&text[len], "%c%d", mark, (int)rint (w));
			else	/* Too small for integer, set as floating point */
				sprintf (&text[len], "%c%lg", mark, w);
```

In english: if the length in PS units exceed 4 then write them as integers, else do floats.  That means we have a max 12.5% round off error for very thin dashed lines (e.g, 0.1p or so - GMT default is 0.25p), and the round off error drops off as lengths increases beyond 4.  Since most of GMT do not use pens < 0.1p it only affects just a few examples.  Perhaps this is OK?
